src/hsprov_store.cpp

```diff
@@ -0,0 +1,256 @@
+const static std::string DIGEST_REALM_COLUMN_NAME    = "digest_realm";
+const static std::string DIGEST_QOP_COLUMN_NAME      = "digest_qop";
+
+// Column name marking rows created by homestead-prov
+const static std::string EXISTS_COLUMN_NAME = "_exists";
+
+// Variables to store the singleton cache object.
+//
+// Must create this after the constants above so that they have been
+// initialized before we initialize the cache.
+HsProvStore* HsProvStore::INSTANCE = &DEFAULT_INSTANCE;
+HsProvStore HsProvStore::DEFAULT_INSTANCE;
+
+//
+// HsProvStore methods
+//
+
+HsProvStore::HsProvStore() : CassandraStore::Store(KEYSPACE) {}
+
+HsProvStore::~HsProvStore() {}
+
+//
+// GetRegData methods
+//
+
+HsProvStore::GetRegData::
+GetRegData(const std::string& public_id) :
+  CassandraStore::HAOperation(),
+  _public_id(public_id),
+  _xml(),
+  _charging_addrs()
+{}
+
+
+HsProvStore::GetRegData::
+~GetRegData()
+{}
+
+
+bool HsProvStore::GetRegData::perform(CassandraStore::Client* client,
+                                SAS::TrailId trail)
+{
+  TRC_DEBUG("Issuing get for key %s", _public_id.c_str());
+
+  std::vector<std::string> requested_columns = {
+    IMS_SUB_XML_COLUMN_NAME,
+    PRIMARY_CCF_COLUMN_NAME,
+    SECONDARY_CCF_COLUMN_NAME,
+    PRIMARY_ECF_COLUMN_NAME,
+    SECONDARY_ECF_COLUMN_NAME
+  };
+
+  std::vector<ColumnOrSuperColumn> results;
+
+  ha_get_columns(client, IMPU, _public_id, requested_columns, results, trail);
+
+  for(std::vector<ColumnOrSuperColumn>::iterator it = results.begin(); it != results.end(); ++it)
+  {
+    if (it->column.name == IMS_SUB_XML_COLUMN_NAME)
+    {
+      _xml = it->column.value;
+      TRC_DEBUG("Retrieved XML column with value %s", _xml.c_str());
+    }
+    else if ((it->column.name == PRIMARY_CCF_COLUMN_NAME) && (it->column.value != ""))
+    {
+      _charging_addrs.ccfs.push_front(it->column.value);
+      TRC_DEBUG("Retrieved primary_ccf column with value %s",
+                it->column.value.c_str());
+    }
+    else if ((it->column.name == SECONDARY_CCF_COLUMN_NAME) && (it->column.value != ""))
+    {
+      _charging_addrs.ccfs.push_back(it->column.value);
+      TRC_DEBUG("Retrieved secondary_ccf column with value %s",
+                it->column.value.c_str());
+    }
+    else if ((it->column.name == PRIMARY_ECF_COLUMN_NAME) && (it->column.value != ""))
+    {
+      _charging_addrs.ecfs.push_front(it->column.value);
+      TRC_DEBUG("Retrieved primary_ecf column with value %s",
+                it->column.value.c_str());
+    }
+    else if ((it->column.name == SECONDARY_ECF_COLUMN_NAME) && (it->column.value != ""))
+    {
+      _charging_addrs.ecfs.push_back(it->column.value);
+      TRC_DEBUG("Retrieved secondary_ecf column with value %s",
+                it->column.value.c_str());
+    }
+  }
+
+  // All exceptions will rise up to the calling function, where the return code
+  // will be set as unsuccessful.
+
+  return true;
```

Comment:
Good question. The method is non-void so it overrides the base class' method. As to why we always return true - that's what we did before.

I think the way this works is that the Operation is allowed to verify the data it has got back, and return a boolean success value, so that we can mark cases as failures when we haven't hit an exception